app.py

```diff
@@ -1,42 +1,99 @@
-        return redirect('/')
-
-    @app.route('/item/add', methods=['POST'])
-    def add_item():
-        db_client.add_item(request.form.get('title'))
-        return redirect('/')
-
-    @app.route('/')
-    def index():
-        items = db_client.get_items()
-        item_view_model = ViewModel(items)
-        return render_template('index.html', view_model=item_view_model)
-
-    return app
+from flask import Flask, render_template, request, redirect
+from config import Config
+from src.status import COMPLETED, IN_PROGRESS
+from src.view_model import ViewModel
+from src.db_client import DbClient, MockDbClient
+from src.oauth import Oauth
+from src.user import User
+from src.role import Role, require_writer
+from flask_login import LoginManager, login_required, login_user, current_user
+
+config = Config('.env')
+
+def create_app(script_info, use_real_db=True, overwrite_db_name=False, mock_login_manager=False):
+    db_client = DbClient(config.db_config, overwrite_db_name) if use_real_db else MockDbClient()
+    
+    app = Flask(__name__)
+    app.config.from_object(config)
+
+    if mock_login_manager:
+        create_mock_login_manager(app)
+    else:
+        create_login_manager(app)
+
+    # Allowing POST so we don't need frontend JS
+    @app.route('/item/<id>/mark-completed', methods=['PUT', 'POST'])
+    @login_required
+    @require_writer
+    def mark_completed(id):
+        db_client.set_status(id, COMPLETED)
+        return redirect('/')
+
+    @app.route('/item/<id>/start-item', methods=['PUT', 'POST'])
+    @login_required
+    @require_writer
+    def start_item(id):
+        db_client.set_status(id, IN_PROGRESS)
+        return redirect('/')
+
+    @app.route('/item/<id>/delete', methods=['DELETE', 'POST'])
+    @login_required
+    @require_writer
+    def delete(id):
+        db_client.delete_item(id)
+        return redirect('/')
+
+    @app.route('/item/add', methods=['POST'])
+    @login_required
+    @require_writer
+    def add_item():
+        db_client.add_item(request.form.get('title'))
+        return redirect('/')
+
+    @app.route('/')
+    @login_required
+    def index():
+        items = db_client.get_items()
+        item_view_model = ViewModel(items, current_user.is_writer)
+        return render_template('index.html', view_model=item_view_model)
+
+    return app
+
+def create_login_manager(app: Flask):
+    login_manager = LoginManager()
+    oauth = Oauth(config.oauth_config)
+
+    @login_manager.unauthorized_handler
+    def unauthenticated():
+        return redirect(oauth.get_oauth_redirect())
+
+    @login_manager.user_loader
+    def load_user(user_id):
+        return User(user_id, Role.for_user(user_id))
+
+    @app.route('/login/oauth-callback')
+    def login():
+        code = request.args.get('code')
+        user = oauth.handleLogin(code)
+        login_user(user)
+        return redirect('/')
+
+    login_manager.init_app(app)
+
+def create_mock_login_manager(app: Flask):

```

Comment:
Does this even really get used? If you set LOGIN_DISABLED=True during tests, then I don't expect these functions to get executed at all.

You instead need to set login_manager.anonymous_user